clients/python-pyo3/src/lib.rs

```diff
@@ -0,0 +1,655 @@
+        parallel_tool_calls: Option<bool>,
+        tags: Option<HashMap<String, String>>,
+        credentials: Option<HashMap<String, ClientSecretString>>,
+    ) -> PyResult<Bound<'a, PyAny>> {
+        let params = BaseTensorZeroGateway::prepare_inference_params(
+            py,
+            function_name,
+            input,
+            episode_id,
+            stream,
+            params,
+            variant_name,
+            dryrun,
+            allowed_tools,
+            additional_tools,
+            tool_choice,
+            parallel_tool_calls,
+            tags,
+            credentials,
+        )?;
+        let client = this.as_super().client.clone();
+        pyo3_async_runtimes::tokio::future_into_py(py, async move {
+            let res = client.inference(params).await;
+            Python::with_gil(|py| match res {
+                Ok(InferenceOutput::NonStreaming(data)) => parse_inference_response(py, data),
+                Ok(InferenceOutput::Streaming(stream)) => Ok(AsyncStreamWrapper {
+                    stream: Arc::new(Mutex::new(stream)),
+                }
+                .into_pyobject(py)?
+                .into_any()
+                .unbind()),
+                Err(e) => Err(convert_error(py, e)?),
+            })
+        })
+    }
+
+    #[pyo3(signature = (*, metric_name, value, inference_id=None, episode_id=None, dryrun=None, tags=None))]
+    fn feedback<'a>(
+        this: PyRef<'a, Self>,
+        metric_name: String,
+        value: Bound<'_, PyAny>,
+        inference_id: Option<Bound<'_, PyAny>>,
+        episode_id: Option<Bound<'_, PyAny>>,
+        dryrun: Option<bool>,
+        tags: Option<HashMap<String, String>>,
+    ) -> PyResult<Bound<'a, PyAny>> {
+        let client = this.as_super().client.clone();
+        let params = BaseTensorZeroGateway::prepare_feedback_params(
+            this.py(),
+            metric_name,
+            value,
+            inference_id,
+            episode_id,
+            dryrun,
+            tags,
+        )?;
+        pyo3_async_runtimes::tokio::future_into_py(this.py(), async move {
+            let res = client.feedback(params).await;
+            Python::with_gil(|py| match res {
+                Ok(resp) => Ok(parse_feedback_response(py, resp)?.into_any()),
+                Err(e) => Err(convert_error(py, e)?),
+            })
+        })
+    }
+}
+
+#[allow(unknown_lints)]
+// This lint currently does nothing on stable, but let's include it
+// so that it will start working automatically when it's stabilized
+#[deny(non_exhaustive_omitted_patterns)]
+fn convert_error(py: Python<'_>, e: TensorZeroError) -> PyResult<PyErr> {
+    match e {
+        TensorZeroError::Http {
+            status_code,
+            text,
+            source: _,
+        } => tensorzero_error(py, status_code, text),
+        TensorZeroError::Other { source } => tensorzero_internal_error(py, &source.to_string()),
+        _ => unreachable!(),
+    }
+}
+
```

Comment:
why do we need this at all? seems like we have exhausted the enum. document the answer for future readers too